toolchain/check/type_structure.cpp

```diff
@@ -0,0 +1,455 @@
+        continue;
+      }
+
+      if (std::holds_alternative<SemIR::SpecificInterface>(next)) {
+        const auto& interface = std::get<SemIR::SpecificInterface>(next);
+        auto args = GetSpecificArgs(interface.specific_id);
+        ++position_;  // Opening of type.
+        if (args.empty()) {
+          structure_.push_back(TypeStructure::Structural::Concrete);
+        } else {
+          structure_.push_back(TypeStructure::Structural::ConcreteOpenParen);
+          Push(CloseType());
+          PushArgs(args);
+        }
+        continue;
+      }
+
+      if (std::holds_alternative<SymbolicType>(next)) {
+        SetFirstSymbolic(position_);
+        ++position_;
+        structure_.push_back(TypeStructure::Structural::Symbolic);
+        continue;
+      }
+
+      if (std::holds_alternative<NonTypeValue>(next)) {
+        ++position_;
+        structure_.push_back(TypeStructure::Structural::Concrete);
+        continue;
+      }
+
+      SemIR::TypeId next_type_id = std::get<SemIR::TypeId>(next);
+      auto inst_id = context_.types().GetInstId(next_type_id);
+      auto inst = context_.insts().Get(inst_id);
+      CARBON_KIND_SWITCH(inst) {
+          // ==== Symbolic types ====
+
+        case CARBON_KIND(SemIR::BindSymbolicName type): {
+          (void)type;
+          Push(SymbolicType());
+          break;
+        }
+        case CARBON_KIND(SemIR::SymbolicBindingPattern type): {
+          (void)type;
+          Push(SymbolicType());
+          break;
+        }
+        case CARBON_KIND(SemIR::FacetAccessType type): {
+          (void)type;
+          Push(SymbolicType());
+          break;
+        }
+
+          // ==== Concrete types ====
+
+        case CARBON_KIND(SemIR::AssociatedEntityType type): {
+          (void)type;
+          ++position_;
+          structure_.push_back(TypeStructure::Structural::Concrete);
+          break;
+        }
+        case CARBON_KIND(SemIR::BoolType type): {
+          (void)type;
+          ++position_;
+          structure_.push_back(TypeStructure::Structural::Concrete);
+          break;
+        }
+        case CARBON_KIND(SemIR::FloatType type): {
+          (void)type;
+          ++position_;
+          structure_.push_back(TypeStructure::Structural::Concrete);
+          break;
+        }
+        case CARBON_KIND(SemIR::GenericClassType type): {
+          (void)type;
+          ++position_;
+          structure_.push_back(TypeStructure::Structural::Concrete);
+          break;
+        }
+        case CARBON_KIND(SemIR::GenericInterfaceType type): {
+          (void)type;
+          ++position_;
+          structure_.push_back(TypeStructure::Structural::Concrete);
+          break;
+        }
+        case CARBON_KIND(SemIR::ImplWitnessAccess type): {
+          (void)type;
+          ++position_;
+          structure_.push_back(TypeStructure::Structural::Concrete);
+          break;
+        }
```

Comment:
Consider merging these identical cases:
```suggestion
        case SemIR::AssociatedEntityType::Kind:
        case SemIR::BoolType::Kind:
        case SemIR::FloatType::Kind:
        case SemIR::GenericClassType::Kind:
        case SemIR::GenericInterfaceType::Kind:
        case SemIR::ImplWitnessAccess::Kind: {
          ++position_;
          structure_.push_back(TypeStructure::Structural::Concrete);
          break;
        }
```
and similar.

Reply:
Oh, I didn't realize I can do that in a CARBON_KIND_SWITCH, thanks!